### PR TITLE
eliminate race, deadlock and dead code

### PR DESCRIPTION
**Pull request by jkt628**

### CODE REVIEW

**Comment by ddragosd** [comment]
thanks @jkt628 . We'll be reviewing this soon.

**Comment by jkt628** [comment]
ping

File: Makefile [hunk 1]
```diff
@@ -21,8 +21,19 @@ static:
 #	CGO_ENABLED=0 GOOS=linux go build -ldflags "-s" -a -installsuffix cgo -o $(GOBIN)/api-gateway-config-supervisor-static ./
 	CGO_ENABLED=0 go build -ldflags "-s" -a -installsuffix cgo -o $(GOBIN)/api-gateway-config-supervisor-static ./
 
-docker:
-	docker build -t adobeapiplatform/api-gateway-config-supervisor .
+DOCKER_TEMP := $(shell mktemp -u ./.Dockerfile.XXXXXXXXXX)
```
Comment (ddragosd) [inline Makefile]: The intent here is to provide a `debug` configuration option when building the docker container ?

File: Dockerfile [hunk 2]
```diff
@@ -6,41 +6,38 @@
...
 
-    && echo "installing rclone sync ... " \
-    && go get github.com/ncw/rclone \
+RUN echo "installing rclone sync ... " \
+    && go get github.com/ncw/rclone
 
-    && echo " cleaning up ... " \
-    && rm -rf /usr/lib/go/bin/src \
-       	      /tmp/go \
+RUN echo " cleaning up ... " \
```
Comment (ddragosd) [inline Dockerfile]: This RUN command to clean up will add a 0 size layer to the docker image but it won't actually reduce its size IIRC. If we want to reduce the size we might have to put all commands into a single RUN instruction.

File: sync/recursiveWatcher.go [hunk 3]
```diff
@@ -13,7 +13,7 @@ import (
 )
 
 type RecursiveWatcher struct {
-	*fsnotify.Watcher
+	fsw     *fsnotify.Watcher
```
Comment (ddragosd) [inline sync/recursiveWatcher.go:16]: was this changed for clarity ? Or it wasn't behaving correctly ?

File: sync/recursiveWatcher.go [hunk 4]
```diff
@@ -76,8 +76,8 @@ func (watcher *RecursiveWatcher) Run(debug bool) {
 						if debug {
 							DebugMessage("Detected new directory %s", event.Name)
 						}
-						if !shouldIgnoreFile(filepath.Base(event.Name)) {
-							watcher.AddFolder(event.Name)
+						for _, folder := range Subfolders(event.Name) {
```
Comment (ddragosd) [inline sync/recursiveWatcher.go:79]: good catch. it would be great to add a unit test for this scenario as well.

**Comment by jkt628** [comment]
no, this is a little preprocessing magic on Dockerfile to speed up development with complex builds.  by default it combines RUN commands into a single layer to decrease container size.  however, define DOCKER_BUILD_DEBUG and those commands go into separate layers which has the possibility to reuse docker cache, saving some time here and there.

this repo was the testbed for apigateway where this change makes a significant difference.

**Comment by ddragosd** [comment]
thanks for clarifying @jkt628 . I very much like your idea of streamlining the local development by replacing subsequent `RUN` commands with `&&`.  I was thinking at extending your idea to communicate the intent a little clearer into the `Dockerfile`. This could be achieved maybe by writing a `SQUASH` instruction above the `RUN` command like in the following example:

``` Dockerfile
RUN echo "I'm the first layer"
#SQUASH
RUN echo "I'm a squashed layer and I belong to the first layer"

RUN echo "I'm a second layer"
```

In this case the final Docker container would have 2 layers. The first 2 `RUN` commands would be squashed into a single layer, and the 3rd `RUN` command would be placed in its own layer.

The question I have next is: how would this optimization work with Docker Hub builds ? Would we actually commit the generated `Dockerfile` as `Dockerfile.optimized` and instruct Docker Hub to build an image using that file instead of the `Dockerfile` ?

**Review by ddragosd — CHANGES_REQUESTED** [review]
@jkt628 would it too much to include 2 extra changes with this PR ?
1. add a unit test for the fix on commit 4cd7591 `when a new directory detected watch the entire subtree`
2. update the README to include the instructions for building the Docker container locally with the DEBUG flag

The other conversation on how to enable the Dockerfile optimization for Docker Hub could be updated on a separate PR. 

Let me know what you think.

**Comment by jkt628** [comment]
happy to oblige in all respects

**Comment by jkt628** [comment]
i thought about the SQUASH request this weekend and think a fully automated approach has greater reach with less effort.  developers generally will not add decorators they neither understand nor trust, assuming they know about it in the first place.  "just do it" applies to more than exercise.

the only good reason for additional layers is if you expect reuse across images which is best expressed by FROM.  this image is not a good candidate.

i welcome debate on this issue, thanks.

File: sync/recursiveWatcher.go (hunk 3 above)
Comment (jkt628) [reply sync/recursiveWatcher.go:16]: this changes RecursiveWatcher's nature from is_a fsnotify.Watcher to has_a fsnotify.Watcher, denying access to fsnotify.Watcher.Events which should prevent future races.

**Comment by ddragosd** [comment]
@jkt628 re `SQUASH` I see your point to make it even simpler for a good cause.  I'm curios what your thoughts are on my other question

> how would this optimization work with Docker Hub builds ? Would we actually commit the generated Dockerfile as Dockerfile.optimized and instruct Docker Hub to build an image using that file instead of the Dockerfile ?

**Comment by jkt628** [comment]
On 2016-10-04 16:00, ddragosd wrote:

> how would this optimization work with Docker Hub builds ? Would we
> actually commit the generated Dockerfile as Dockerfile.optimized and
> instruct Docker Hub to build an image using that file instead of the
> Dockerfile ?
1. i am fairly new to the game so maybe i don't understand the process 
   fully but i thought Docker Hub merely accepts your already built images 
   and am not aware that it builds images itself.
2. the extant Dockerfile builds images, just not optimized.  the README 
   already suggests our preferred method and that builds optimized images. 
   i think that is sufficient.

:{)}

**Comment by ddragosd** [comment]
yes, one can build the image separately and then push it to Docker Hub.

Another easier way is to let Docker Hub automatically build the docker image on every git-push.
The UI looks like this: 
image

This is the case I was talking about in which the repo has to have an optimized dockerfile somewhere.

**Comment by jkt628** [comment]
i played with Docker Hub Automated Build and agree mine is not a good solution as Dockerfile is essentially a reserved word.  i will submit optimized Dockerfile for this PR and the upcoming PR for apigateway and later design a different approach for developers.

**Comment by jkt628** [comment]
@ddragosd these commits should address these issues:
1. add a unit test for the fix on commit 4cd7591 when a new directory detected watch the entire subtree
2. Another easier way is to let Docker Hub automatically build the docker image on every git-push

this issue no longer applies:
1. update the README to include the instructions for building the Docker container locally with the DEBUG flag

**Comment by jkt628** [comment]
this merge requires a new tag so i can submit a PR for [apigateway](https://github.com/adobe-apiplatform/apigateway)

**Comment by ddragosd** [comment]
thanks @jkt628 . I'm checking the changes right now. 

> this merge requires a new tag so i can submit a PR for apigateway

I can help with making a new release as soon as we merge these changes.

BTW, we can also use commit SHAs to download tarballs even if we don't have a tag. i.e.
 https://github.com/adobe-apiplatform/api-gateway-config-supervisor/archive/05fcbb5.tar.gz

This might help with local testing once we commit something on a branch.

**Review by ddragosd — APPROVED** [review]

**Comment by ddragosd** [comment]
👍

**Comment by ddragosd** [comment]
@jkt628 I've drafted a new release candidate here: https://github.com/adobe-apiplatform/api-gateway-config-supervisor/releases/tag/1.0.1-RC1

**Comment by jkt628** [comment]
thanks for the quick response @ddragosd .  i think api-gateway-config-supervisor is not used by itself so will test via updates to apigateway.

**Pull request merged**